vectors/no_data_vec.rs

```diff
@@ -1,215 +1,114 @@
-            if self.capacity == self.len {
-                self.reserve(1);
-            }
+    pub fn push(&mut self, _value: T) {
+        if self.len == self.capacity {
+            self.grow(1);
         }
-
-        assert!(self.capacity >= self.len);
-        // We only increment the length of the vector disregarding the actual
-        // element added to the Vector.
         self.len += 1;
+        assert!(self.capacity >= self.len);
     }
 
-    // We check if there are any elements in the Vector. If not, we return a None
-    // otherwise we return a nondeterministic value since we dont track any concrete
-    // values in the Vector.
-    #[cfg_attr(feature = "certora-debug", inline(never))]
     pub fn pop(&mut self) -> Option<T> {
         if self.len == 0 {
             None
         } else {
             self.len -= 1;
-	    Some(nondet::<T>())
+            Some(nondet::<T>())
         }
     }
 
-    #[cfg_attr(feature = "certora-debug", inline(never))]
-    pub fn append(&mut self, other: &mut NoDataVec<T>) {
-        let new_len = self.len + other.len;
-
-        if allow_resizing!() {
-            // Please refer to grow() for better understanding the semantics of grow().
-            if self.capacity < new_len {
-                self.reserve(other.len);
-            }
+    pub fn insert(&mut self, _index: usize, _value: T) {
+        if self.len == self.capacity {
+            self.grow(1);
         }
-
-        assert!(self.capacity >= new_len);
-        // Drop all writes, increment the length of the Vector with the size
-        // of the Vector which is appended.
-        self.len = new_len;
-    }
-
-    // At whichever position we insert the new element into, the overall effect on
-    // the abstraction is that the length increases by 1.
-    #[cfg_attr(feature = "certora-debug", inline(never))]
-    pub fn insert(&mut self, _index: usize, _elem: T) {
-        assert!(index <= self.len);
-
         self.len += 1;
+        assert!(self.capacity >= self.len);
     }
 
-    // We only care that the index we are removing from lies somewhere as part of
-    // the length of the Vector. The only effect on the abstraction is that the
-    // length decreases by 1. In the case that it is a valid removal, we return a
-    // nondeterministic value.
-    #[cfg_attr(feature = "certora-debug", inline(never))]
     pub fn remove(&mut self, _index: usize) -> T {
         assert!(index < self.len);
-
         self.len -= 1;
-	    nondet::<T>()
+        nondet::<T>()
     }
 
-    #[cfg_attr(feature = "certora-debug", inline(never))]
-    pub fn extend<I: Iterator>(&mut self, iter: I)
-    where
-        I: Iterator<Item = T>,
-    {
-        // We first compute the length of the iterator.
-        let mut iter_len = 0;
-        for _value in iter {
-            iter_len += 1;
+    pub fn get(&self, _index: usize) -> Option<&T> {
+        assert!( index < self.len);
+        let ptr: *mut T = nondet_pointer::<T>();
```

Comment:
is this necessary? 

If yes, we want a different implementation. All pointers should be accounted for. Nondet pointer is a very difficult to support since it can point anywhere. This is more of a fresh pointer somewhere. For example, we can allocate one element of type `T` and always return its address.

Reply:
This is @caballa 's comment on `nondet pointer`:
```
/// The returned pointer is a raw pointer and we won't free it.
/// That's okay since this function is only for verification purposes.
/// That's why we alloc a raw pointer to avoid including all the ownership/reference counting
/// stuff from Box and Rc
pub fn nondet_pointer<T: Nondet>() -> *mut T {
    unsafe {
        let layout = Layout::new::<T>();
        let ptr = alloc(layout) as *mut T;
        CVT_assume(!ptr.is_null());
        *ptr = nondet();
        ptr
    }
}
```
Would this still be an issue?

Reply:
Now what I imagined `nondet_pointer` to be. This is good. It is a fresh allocated pointer into a memory cell of type `T`. This is what I wanted to see.